File: internal/ui/livetable.go
package ui

import (
	"fmt"
	"time"

	"github.com/charmbracelet/bubbles/spinner"
	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
	"github.com/charmbracelet/lipgloss/table"
	"github.com/claby2/hladmin/internal/executor"
)

// TableSpec describes how to turn per-host execution state into table rows. The
// caller owns the column semantics and cell styling; this package owns layout and
// animation.
type TableSpec struct {
	Headers []string
	// CompletedRow returns the cells for a finished host.
	CompletedRow func(executor.Result) []string
	// RunningRow returns the cells for a host still in progress. spinnerFrame is
	// the current spinner glyph.
	RunningRow func(host, spinnerFrame string, elapsed time.Duration) []string
}

// renderTable lays out headers and rows with a compact, borderless style. When
// maxWidth > 0 and the natural table is wider, it is constrained so no line
// exceeds the terminal width.
func renderTable(spec TableSpec, rows [][]string, maxWidth int) string {
	lastCol := len(spec.Headers) - 1
	padded := lipgloss.NewStyle().PaddingRight(2)
	plain := lipgloss.NewStyle()
	t := table.New().
		Border(lipgloss.HiddenBorder()).
		BorderTop(false).BorderBottom(false).BorderLeft(false).BorderRight(false).
		BorderColumn(false).BorderRow(false).BorderHeader(false).
		StyleFunc(func(row, col int) lipgloss.Style {
			if col == lastCol {
				return plain
			}
			return padded
		}).
		Headers(spec.Headers...).
		Rows(rows...)

	if maxWidth > 0 && lipgloss.Width(t.String()) > maxWidth {
		t = t.Width(maxWidth)
	}
	return t.String()
}

// tableModel renders a live, in-place table that fills in as hosts complete.
type tableModel struct {
	hosts     []string
	command   string
	spec      TableSpec
	spinner   spinner.Model
	starts    []time.Time
	done      []bool
	results   []executor.Result
	remaining int
	width     int
}

func (m tableModel) Init() tea.Cmd {
	cmds := []tea.Cmd{m.spinner.Tick}
	for i, host := range m.hosts {
		cmds = append(cmds, runHostCmd(i, host, m.command))
	}
	return tea.Batch(cmds...)
}

func (m tableModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	switch msg := msg.(type) {
	case hostDoneMsg:
		m.done[msg.index] = true
		m.results[msg.index] = msg.result
		m.remaining--
		if m.remaining == 0 {
			return m, tea.Quit
		}
		return m, nil
	case spinner.TickMsg:
		var cmd tea.Cmd
		m.spinner, cmd = m.spinner.Update(msg)
		return m, cmd
	case tea.WindowSizeMsg:
		m.width = msg.Width
		return m, nil
	case tea.KeyMsg:
		if msg.Type == tea.KeyCtrlC {
			return m, tea.Quit
		}
	}
	return m, nil
}

func (m tableModel) rows() [][]string {
	rows := make([][]string, len(m.hosts))
	for i, host := range m.hosts {
		if m.done[i] {
			rows[i] = m.spec.CompletedRow(m.results[i])
		} else {
			rows[i] = m.spec.RunningRow(host, m.spinner.View(), time.Since(m.starts[i]))
		}
	}
	return rows
}

func (m tableModel) View() string {
	return renderTable(m.spec, m.rows(), m.width) + "\n"
}

// RunLiveTable executes command on hosts in parallel and renders a live table via
// spec, redrawing in place as hosts complete (TTY). On a non-TTY it renders the
// final table once after all hosts finish.
func RunLiveTable(hosts []string, command string, spec TableSpec) ([]executor.Result, error) {
	if err := executor.VerifyHostsAndCommand(hosts, command); err != nil {
		return nil, err
	}

	if !IsTerminal() {
		results := runAllParallel(hosts, command)
		rows := make([][]string, len(results))
		for i, r := range results {
			rows[i] = spec.CompletedRow(r)
		}
		fmt.Println(renderTable(spec, rows, 0))
		return results, nil
	}

	now := time.Now()
	starts := make([]time.Time, len(hosts))
	for i := range starts {
		starts[i] = now
	}

	m := tableModel{
		hosts:     hosts,
		command:   command,
		spec:      spec,
		spinner:   newSpinner(),
		starts:    starts,
		done:      make([]bool, len(hosts)),
		results:   make([]executor.Result, len(hosts)),
		remaining: len(hosts),
	}

	final, err := tea.NewProgram(m).Run()
	if err != nil {
		return nil, err
	}
	return final.(tableModel).results, nil
}

File: src/ui/livetable.rs
use crate::executor::{self, ExecResult};
use crate::ui::stream::{SPINNER_FRAMES, TICK_INTERVAL, collect_results};
use crate::ui::styles;
use anyhow::Result;
use comfy_table::{ContentArrangement, Table};
use indicatif::{ProgressBar, ProgressDrawTarget, ProgressStyle};
use std::sync::mpsc::RecvTimeoutError;
use std::time::{Duration, Instant};

/// Describes how to turn per-host execution state into table rows. The caller
/// owns the column semantics and cell styling; this module owns layout and
/// animation.
pub struct TableSpec {
    pub headers: Vec<&'static str>,
    /// Returns the cells for a finished host.
    pub completed_row: CompletedRowFn,
    /// Returns the cells for a host still in progress. `spinner_frame` is the
    /// current spinner glyph.
    pub running_row: RunningRowFn,
}

pub type CompletedRowFn = Box<dyn Fn(&ExecResult) -> Vec<String>>;
pub type RunningRowFn = Box<dyn Fn(&str, &str, Duration) -> Vec<String>>;

/// Lays out headers and rows with a compact, borderless style. When max_width
/// is given, the table is constrained so no line exceeds the terminal width.
fn render_table(spec: &TableSpec, rows: Vec<Vec<String>>, max_width: Option<u16>) -> String {
    let mut table = Table::new();
    table.load_preset(comfy_table::presets::NOTHING);
    table.set_header(spec.headers.clone());
    for row in rows {
        table.add_row(row);
    }

    let last = spec.headers.len() - 1;
    for (i, column) in table.column_iter_mut().enumerate() {
        column.set_padding((0, if i == last { 0 } else { 2 }));
    }

    if let Some(width) = max_width {
        // Dynamic arrangement keeps natural column widths and only shrinks when
        // the table would exceed the terminal width.
        table.set_content_arrangement(ContentArrangement::Dynamic);
        table.set_width(width);
    }

    table.to_string()
}

/// Executes command on hosts in parallel and renders a live table via spec,
/// redrawing in place as hosts complete (TTY). On a non-TTY it renders the
/// final table once after all hosts finish.
pub fn run_live_table(hosts: &[String], command: &str, spec: TableSpec) -> Result<Vec<ExecResult>> {
    executor::verify_hosts_and_command(hosts, command)?;

    let start = Instant::now();
    let rx = executor::run_parallel(hosts, command);

    if !styles::is_terminal() {
        let mut results: Vec<Option<ExecResult>> = hosts.iter().map(|_| None).collect();
        for _ in hosts {
            let Ok((i, result)) = rx.recv() else { break };
            results[i] = Some(result);
        }
        let results = collect_results(results, hosts, command, start);
        let rows = results.iter().map(|r| (spec.completed_row)(r)).collect();
        println!("{}", render_table(&spec, rows, None));
        return Ok(results);
    }

    // A single spinner bar used purely as a multi-line redraw region; indicatif
    // owns the cursor movement and line clearing.
    let pb = ProgressBar::with_draw_target(None, ProgressDrawTarget::stdout());
    pb.set_style(ProgressStyle::with_template("{msg}").expect("valid template"));

    let term_width = console::Term::stdout().size().1;
    let mut results: Vec<Option<ExecResult>> = hosts.iter().map(|_| None).collect();
    let mut remaining = hosts.len();
    let mut frame = 0usize;

    let build_rows = |results: &[Option<ExecResult>], frame: usize| -> Vec<Vec<String>> {
        results
            .iter()
            .enumerate()
            .map(|(i, result)| match result {
                Some(r) => (spec.completed_row)(r),
                None => (spec.running_row)(&hosts[i], SPINNER_FRAMES[frame], start.elapsed()),
            })
            .collect()
    };

    pb.set_message(render_table(
        &spec,
        build_rows(&results, frame),
        Some(term_width),
    ));
    while remaining > 0 {
        match rx.recv_timeout(TICK_INTERVAL) {
            Ok((i, result)) => {
                results[i] = Some(result);
                remaining -= 1;
            }
            Err(RecvTimeoutError::Timeout) => {
                frame = (frame + 1) % SPINNER_FRAMES.len();
            }
            Err(RecvTimeoutError::Disconnected) => break,
        }
        pb.set_message(render_table(
            &spec,
            build_rows(&results, frame),
            Some(term_width),
        ));
    }

    pb.finish_and_clear();
    let results = collect_results(results, hosts, command, start);
    let rows = results.iter().map(|r| (spec.completed_row)(r)).collect();
    println!("{}", render_table(&spec, rows, Some(term_width)));
    Ok(results)
}
